dmscripts/notify_buyers_to_award_closed_briefs.py

```diff
@@ -0,0 +1,144 @@
+# -*- coding: utf-8 -*-
+from datetime import datetime, date, timedelta
+
+import dmapiclient
+from dmutils.email import DMNotifyClient
+from dmutils.email.exceptions import EmailError
+from dmutils.formats import DATE_FORMAT
+
+from dmscripts.helpers import logging_helpers, date_helpers, brief_data_helpers
+
+logger = logging_helpers.configure_logger({'dmapiclient': logging_helpers.logging.INFO})
+
+
+def _filter_brief_users_by_optional_user_id_list(brief, user_id_list):
+    """Filter the brief's users by restricted user ID list if present"""
+    if user_id_list:
+        return filter(lambda user: user['id'] in user_id_list, brief['users'])
+    else:
+        return brief['users']
+
+
+def _filter_briefs_by_user_id_list(briefs, user_id_list):
+    """Return a list of briefs if they have at least 1 user in the given ID list"""
+    return [brief for brief in briefs if any(_filter_brief_users_by_optional_user_id_list(brief, user_id_list))]
+
+
+def _create_context_for_brief(brief):
+    return {
+        'brief_id': brief['id'],
+        'brief_title': brief['title'],
+        'utm_date': date.today().strftime("%Y%m%d")
+    }
+
+
+def notify_users(notify_client, notify_template_id, brief, user_id_list):
+    failed_users = []
+
+    context_data = _create_context_for_brief(brief)
+    brief_users = _filter_brief_users_by_optional_user_id_list(brief, user_id_list)
+
+    for user in brief_users:
+        if user['active']:
+            try:
+                notify_client.send_email(
```

Comment:
Really really sorry, I should have mentioned/spotted this before, but...

I think *this* is the call that should be wrapped with an `if dry_run...`, and not the higher-level call to `notify_users`.  As things stand the `dry-run` will not test the creation of context data or filtering users.  You want the dry run to run as much of the script as possible, and only skip the API calls that update data/send an email or whatever to give the fullest possible test of what would happen, just without doing the actual thing.

Sorry.

Reply:
No that makes perfect sense! We were trying to keep all the logging at the same depth, so the flag got bumped up into `main`. I will have a look and see if I can get it closer to the `notify_client` call.